lib/Target/Btor/BtorToBtorIRTranslation.cpp

```diff
@@ -0,0 +1,517 @@
+      testResults[i] = cache.at(inits.at(j++)->args[1]);
+    } else {
+      auto sort = returnTypes.at(i).getIntOrFloatBitWidth();
+      if (undefOpsBySort.find(sort) == undefOpsBySort.end()) {
+          auto res = builder.create<btor::UndefOp>(unknownLoc,
+                                    returnTypes.at(i));
+          assert(res && res->getNumResults() == 1);
+          undefOpsBySort[sort] = res->getResult(0);
+      }
+      testResults[i] = undefOpsBySort.at(sort);
+    }
+    assert(testResults[i]);
+  }
+
+  builder.create<ReturnOp>(unknownLoc, testResults);
+
+  return funcOp;
+}
+
+OwningOpRef<FuncOp> Deserialize::buildNextFunction() {
+  // collect the return types for our init function
+  std::vector<Type> returnTypes(nexts.size(), nullptr);
+  for (uint32_t i = 0; i < nexts.size(); ++i) {
+    returnTypes[i] = builder.getIntegerType(nexts.at(i)->sort.bitvec.width);
+    assert(returnTypes[i]);
+  }
+
+  // create next function signature
+  OperationState state(unknownLoc, FuncOp::getOperationName());
+  FuncOp::build(builder, state, "next",
+                FunctionType::get(context, returnTypes, returnTypes));
+  OwningOpRef<FuncOp> funcOp = cast<FuncOp>(Operation::create(state));
+  Region &region = funcOp->getBody();
+  OpBuilder::InsertionGuard guard(builder);
+  auto *body = builder.createBlock(&region, {}, returnTypes);
+  builder.setInsertionPointToStart(body);
+
+  // clear cache so that values are mapped to the right Basic Block
+  cache.clear();
+  // initialize states with block arguments
+  for (uint32_t i = 0; i < nexts.size(); ++i) {
+    cache[nexts.at(i)->args[0]] = body->getArguments()[i];
+  }
+
+  // start with nexts, then add bads, for logic sharing
+  for (auto next : nexts) { toOp(next); }
+  for (auto bad : bads) { toOp(bad); }
+
+  // close with a fitting returnOp
+  std::vector<Value> testResults(nexts.size(), nullptr);
+  for (uint32_t i = 0; i < nexts.size(); ++i) {
+    testResults[i] = cache.at(nexts.at(i)->args[1]);
+    assert(testResults[i]);
+  }
+
+  builder.create<ReturnOp>(unknownLoc, testResults);
+  return funcOp;
+}
+
+static OwningModuleRef deserializeModule(const llvm::MemoryBuffer *input,
+                                         MLIRContext *context) {
+  context->loadDialect<btor::BtorDialect>();
+  context->loadDialect<StandardOpsDialect>();
+
+  OwningModuleRef owningModule(ModuleOp::create(FileLineColLoc::get(
+      context, input->getBufferIdentifier(), /*line=*/0, /*column=*/0)));
+
+  Deserialize deserialize(context, input->getBufferIdentifier().str());
+  if (deserialize.parseModelIsSuccessful()) {
+    OwningOpRef<FuncOp> initFunc = deserialize.buildInitFunction();
+    if (!initFunc)
+      return {};
+
+    OwningOpRef<FuncOp> nextFunc = deserialize.buildNextFunction();
+    if (!nextFunc)
+      return {};
+
+    owningModule->getBody()->push_front(nextFunc.release());
+    owningModule->getBody()->push_front(initFunc.release());
+  }
+
+  return owningModule;
```

Comment:
seems like this is returning a null pointer on failure some times and an empty module some other times. Unify error behavior